File: src/main/java/org/jitsi/jicofo/health/JicofoHealthChecker.java
/*
 * Copyright @ 2018 - present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jitsi.jicofo.health;

import kotlin.*;
import org.jitsi.health.*;
import org.jitsi.impl.protocol.xmpp.*;
import org.jitsi.jicofo.*;
import org.jitsi.jicofo.bridge.*;
import org.jitsi.jicofo.xmpp.*;
import org.jitsi.utils.logging2.*;
import org.jitsi.utils.logging2.Logger;
import org.jivesoftware.smack.*;
import org.jivesoftware.smackx.ping.packet.*;
import org.jxmpp.jid.*;
import org.jxmpp.jid.impl.*;
import org.jxmpp.jid.parts.*;
import org.jxmpp.stringprep.*;

import java.time.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.logging.*;

import static org.jitsi.jicofo.health.HealthConfig.config;

/**
 * Checks the health of {@link FocusManager}.
 *
 * @author Lyubomir Marinov
 * @author Pawel Domas
 */
public class JicofoHealthChecker implements HealthCheckService
{
    /**
     * The {@code Logger} utilized by the {@code Health} class to print
     * debug-related information.
     */
    private static final Logger logger = new LoggerImpl(JicofoHealthChecker.class.getName());

    /**
     * The {@code JitsiMeetConfig} properties to be utilized for the purposes of
     * checking the health (status) of Jicofo.
     */
    private static final Map<String,String> JITSI_MEET_CONFIG = Collections.emptyMap();

    /**
     * The pseudo-random generator used to generate random input for
     * {@link FocusManager} such as room names.
     */
    private static final Random RANDOM = new Random();

    /**
     * Counts how many health checks took too long.
     */
    private long totalSlowHealthChecks = 0;

    private FocusManager focusManager;
    private final HealthChecker healthChecker;

    public JicofoHealthChecker(HealthConfig config, FocusManager focusManager)
    {
        this.focusManager = focusManager;
        this.healthChecker = new HealthChecker(
                config.getInterval(),
                config.getTimeout(),
                config.getMaxCheckDuration(),
                false,
                Duration.ofMinutes(5),
                this::performCheck,
                Clock.systemUTC());

    }

    public void start()
    {
        healthChecker.start();
    }

    public void stop()
    {
        focusManager = null;
        try
        {
            healthChecker.stop();
        }
        catch (Exception e)
        {
            logger.warn("Failed to stop.", e);
        }
    }

    public Unit performCheck()
    {
        Objects.requireNonNull(focusManager, "FocusManager is not set.");

        long start = System.currentTimeMillis();

        check(focusManager);

        long duration = System.currentTimeMillis() - start;

        if (duration > config.getMaxCheckDuration().toMillis())
        {
            logger.error("Health check took too long: " + duration + "ms");
            totalSlowHealthChecks++;
        }

        return Unit.INSTANCE;
    }

    /**
     * Checks the health (status) of a specific {@link FocusManager}.
     *
     * @param focusManager the {@code FocusManager} to check the health (status)
     * of
     * @throws Exception if an error occurs while checking the health (status)
     * of {@code focusManager} or the check determines that {@code focusManager}
     * is not healthy
     */
    private static void check(FocusManager focusManager)
    {
        // Get the MUC service to perform the check on.
        JicofoServices jicofoServices = JicofoServices.jicofoServicesSingleton;
        if (jicofoServices == null)
        {
            throw new RuntimeException("No JicofoServices available");
        }

        BridgeSelector bridgeSelector = jicofoServices.getBridgeSelector();
        if (bridgeSelector.getOperationalBridgeCount() <= 0)
        {
            throw new RuntimeException(
                    "No operational bridges available (total bridge count: "
                            + bridgeSelector.getBridgeCount() + ")");
        }

        // Generate a pseudo-random room name. Minimize the risk of clashing
        // with existing conferences.
        EntityBareJid roomName;

        do
        {
            roomName = JidCreate.entityBareFrom(
                generateRoomName(),
                XmppConfig.client.getConferenceMucJid()
            );
        }
        while (focusManager.getConference(roomName) != null);

        // Create a conference with the generated room name.
        try
        {
            if (!focusManager.conferenceRequest(
                    roomName,
                    JITSI_MEET_CONFIG,
                    Level.WARNING /* conference logging level */,
                    false /* don't include in statistics */))
            {
                throw new RuntimeException("Failed to create conference with room name " + roomName);
            }

            pingClientConnection();
        }
        catch (Exception e)
        {
            throw new RuntimeException("Failed to create conference with room name " + roomName + ":" + e.getMessage());
        }
    }

    /**
     * The check focusManager.conferenceRequest uses Smack's collectors for the response which is executed
     * in its Reader thread. This ping test use a sync stanza listener which is executed in a single thread
     * and if something is blocking it healthcheck will fail.
     *
     * @throws Exception if the check fails or some other error occurs
     */
    private static void pingClientConnection()
        throws Exception
    {
        CountDownLatch pingResponseWait = new CountDownLatch(1);
        Ping p = new Ping(JidCreate.bareFrom(XmppConfig.client.getXmppDomain()));

        XmppProvider provider = Objects.requireNonNull(JicofoServices.jicofoServicesSingleton).getXmppServices()
            .getClientConnection();
        XMPPConnection connection = provider.getXmppConnection();
        StanzaListener listener = packet -> pingResponseWait.countDown();
        try
        {
            connection.addSyncStanzaListener(
                listener, stanza -> stanza.getStanzaId() != null && stanza.getStanzaId().equals(p.getStanzaId())
            );
            connection.sendStanza(p);

            // will wait for 5 seconds to receive the ping
            if (!pingResponseWait.await(5, TimeUnit.SECONDS))
            {
                throw new RuntimeException("did not receive ping from the xmpp server");
            }
        }
        finally
        {
            connection.removeSyncStanzaListener(listener);
        }
    }

    /**
     * Generates a pseudo-random room name which is not guaranteed to be unique.
     *
     * @return a pseudo-random room name which is not guaranteed to be unique
     */
    private static Localpart generateRoomName()
    {
        try
        {
            return
                Localpart.from(config.getRoomNamePrefix()
                    + "-"
                    + Long.toHexString(System.currentTimeMillis() + RANDOM.nextLong()));
        }
        catch (XmppStringprepException e)
        {
            // ignore, cannot happen
            return null;
        }
    }

    /**
     * @return how many health checks took too long so far.
     */
    public long getTotalSlowHealthChecks()
    {
        return totalSlowHealthChecks;
    }

    @Override
    public Exception getResult()
    {
        return healthChecker.getResult();
    }
}

File: jicofo/src/main/kotlin/org/jitsi/jicofo/health/JicofoHealthChecker.kt
/*
 * Copyright @ 2018 - present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jitsi.jicofo.health

import org.jitsi.health.HealthCheckService
import org.jitsi.health.HealthChecker
import org.jitsi.health.Result
import org.jitsi.jicofo.FocusManager
import org.jitsi.jicofo.bridge.BridgeSelector
import org.jitsi.jicofo.metrics.JicofoMetricsContainer
import org.jitsi.jicofo.xmpp.XmppConfig
import org.jitsi.jicofo.xmpp.XmppProvider
import org.jitsi.utils.logging2.createLogger
import org.jivesoftware.smack.SmackException
import org.jivesoftware.smack.StanzaListener
import org.jivesoftware.smackx.ping.packet.Ping
import org.jxmpp.jid.EntityBareJid
import org.jxmpp.jid.impl.JidCreate
import org.jxmpp.jid.parts.Localpart
import java.time.Clock
import java.time.Duration
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.logging.Level
import kotlin.random.Random

/**
 * Checks the health of Jicofo.
 *
 * @author Lyubomir Marinov
 * @author Pawel Domas
 */
class JicofoHealthChecker(
    config: HealthConfig,
    private val focusManager: FocusManager,
    private val bridgeSelector: BridgeSelector,
    private val xmppProviders: Collection<XmppProvider>
) : HealthCheckService {
    private val logger = createLogger()

    /** Counts how many health checks took too long. */
    var totalSlowHealthChecks: Long = 0
        private set

    private val healthChecker = HealthChecker(
        config.interval,
        config.timeout,
        config.maxCheckDuration,
        false,
        Duration.ofMinutes(5),
        { performCheck() },
        Clock.systemUTC()
    )

    fun start() = healthChecker.start()

    fun shutdown() {
        try {
            healthChecker.stop()
        } catch (e: Exception) {
            logger.warn("Failed to stop.", e)
        }
    }

    private fun performCheck(): Result {
        val start = System.currentTimeMillis()
        return check().also {
            val duration = System.currentTimeMillis() - start
            if (duration > HealthConfig.config.maxCheckDuration.toMillis()) {
                logger.error("Health check took too long: $duration ms")
                totalSlowHealthChecks++
            }
            healthyMetric.set(it.success)
        }
    }

    override val result: Result
        get() = healthChecker.result

    /**
     * Checks the health (status) of a specific [FocusManager].
     *
     * @throws Exception if an error occurs while checking the health (status)
     * of `focusManager` or the check determines that `focusManager`
     * is not healthy
     */
    private fun check(): Result {
        if (bridgeSelector.operationalBridgeCount <= 0) {
            return Result(
                success = false,
                hardFailure = true,
                message = "No operational bridges available (total bridge count: ${BridgeSelector.bridgeCount.get()})"
            )
        }

        // Generate a pseudo-random room name. Minimize the risk of clashing with existing conferences.
        var roomName: EntityBareJid
        do {
            roomName = JidCreate.entityBareFrom(
                generateRoomName(),
                XmppConfig.client.conferenceMucJid
            )
        } while (focusManager.getConference(roomName) != null)

        // Create a conference with the generated room name.
        try {
            if (!focusManager.conferenceRequest(
                    roomName,
                    emptyMap(),
                    loggingLevel = Level.WARNING,
                    includeInStatistics = false
                ).isStarted
            ) {
                return Result(success = false, hardFailure = true, message = "Test conference failed to start.")
            }
        } catch (e: SmackException.NoResponseException) {
            // Treat timeouts as "soft" as they could be just due to load.
            return Result(
                success = false,
                hardFailure = false,
                message = "Test conference failed to start due to timeout."
            )
        }

        try {
            xmppProviders.forEach { pingXmppProvider(it) }
        } catch (e: Exception) {
            return Result(success = false, hardFailure = false, message = "XMPP ping failed: ${e.message}")
        }

        return Result(success = true)
    }

    /**
     * The check focusManager.conferenceRequest uses Smack's collectors for the response which is executed
     * in its Reader thread. This ping test use a sync stanza listener which is executed in a single thread
     * and if something is blocking it healthcheck will fail.
     *
     * @throws Exception if the check fails or some other error occurs
     */
    @Throws(Exception::class)
    private fun pingXmppProvider(xmppProvider: XmppProvider) {
        val pingResponseWait = CountDownLatch(1)
        val xmppDomain = xmppProvider.config.xmppDomain
        if (xmppDomain == null) {
            logger.debug("Not pinging ${xmppProvider.config.name}, domain not configured.")
            return
        }

        val p = Ping(JidCreate.bareFrom(xmppDomain))
        val listener = StanzaListener { _ -> pingResponseWait.countDown() }
        try {
            xmppProvider.xmppConnection.addSyncStanzaListener(listener) { it.stanzaId == p.stanzaId }
            xmppProvider.xmppConnection.sendStanza(p)

            // Wait for 5 seconds to receive a response.
            if (!pingResponseWait.await(5, TimeUnit.SECONDS)) {
                throw RuntimeException("did not receive xmpp ping response for ${xmppProvider.config.name}")
            }
        } finally {
            xmppProvider.xmppConnection.removeSyncStanzaListener(listener)
        }
    }

    companion object {
        val healthyMetric = JicofoMetricsContainer.instance.registerBooleanMetric(
            "healthy",
            "Whether jicofo is healthy or not.",
            true
        )
    }
}

private fun generateRoomName(): Localpart = Localpart.from("${HealthConfig.config.roomNamePrefix}-${Random.nextLong()}")
